code/modules/mob/living/carbon/human/human_update_icons.dm

```diff
@@ -573,58 +573,99 @@ There are several things that need to be remembered:
-// Used in approximating what color the pants of clothing should be
+/// Modifies a sprite to conform to digitigrade body shapes
+/proc/wear_digi_version(icon/base_icon, obj/item/item, key, greyscale_colors)
+	ASSERT(istype(item), "wear_digi_version: no item passed")
+	ASSERT(istext(key), "wear_digi_version: no key passed")
+	if(isnull(greyscale_colors) || length(SSgreyscale.ParseColorString(greyscale_colors)) > 1)
+		greyscale_colors = item.get_general_color(base_icon)
+
+	var/index = "[key]-[item.type]-[greyscale_colors]"
+	var/static/list/digitigrade_clothing_cache = list()
+	var/icon/resulting_icon = digitigrade_clothing_cache[index]
+	if(!resulting_icon)
+		var/digi_icon_state
+		if(istype(item, /obj/item/clothing/under))
+			digi_icon_state = "jumpsuit_worn"
+		else if(istype(item, /obj/item/clothing/suit))
+			digi_icon_state = "oversuit_worn"
+		else if(istype(item, /obj/item/clothing/shoes/sneakers))
+			digi_icon_state = "sneakers_worn"
+		else if(istype(item, /obj/item/clothing/shoes))
+			digi_icon_state = "boots_worn"
+
+		if(!digi_icon_state)
+			stack_trace("wear_digi_version: item type [item.type] does not have a digitigrade icon state")
+			return base_icon
+
+		resulting_icon = icon(SSgreyscale.GetColoredIconByType(/datum/greyscale_config/digitigrade, greyscale_colors), digi_icon_state)
+
+		if(item.body_parts_covered & LEGS)
+			var/static/icon/leg_mask
+			if(!leg_mask)
+				leg_mask = icon('icons/mob/clothing/under/masking_helpers.dmi', "digi_leg_mask")
+
+			// cuts the legs off
+			base_icon.Blend(leg_mask, ICON_SUBTRACT)
+			// staples the new legs on
+			base_icon.Blend(resulting_icon, ICON_OVERLAY)
+			resulting_icon = base_icon
+
+		digitigrade_clothing_cache[index] = fcopy_rsc(resulting_icon)
+
+	return icon(resulting_icon)
+
+/**
+ * Get what color the item is on "average"
+ * Can be used to approximate what color this item is/should be
+ *
+ * Arguments:
+ * * base_icon: The icon to get the color from
+ */
+/obj/item/proc/get_general_color(icon/base_icon)
+	if(greyscale_colors && length(SSgreyscale.ParseColorString(greyscale_colors)) == 1)
+		return greyscale_colors
+	return color
+
+// These coordinates point to the middle of the left leg
 #define LEG_SAMPLE_X_LOWER 13
 #define LEG_SAMPLE_X_UPPER 14
-
 #define LEG_SAMPLE_Y_LOWER 8
 #define LEG_SAMPLE_Y_UPPER 9
 
-/// Modifies a sprite to conform to digitigrade body shapes
-/proc/wear_digi_version(icon/base_icon, key, greyscale_config = /datum/greyscale_config/jumpsuit/worn_digi, greyscale_colors)
-	ASSERT(key, "wear_digi_version: no key passed")
-	ASSERT(ispath(greyscale_config, /datum/greyscale_config), "wear_digi_version: greyscale_config is not a valid path (got: [greyscale_config])")
-	// items with greyscale colors containing multiple colors are invalid
-	if(isnull(greyscale_colors) || length(SSgreyscale.ParseColorString(greyscale_colors)) > 1)
+/obj/item/clothing/get_general_color(icon/base_icon)
+	if(body_parts_covered & LEGS)
```

Comment:
I'd return this to check slot flags rather than bodyparts covered, it's more bespoke and less prone to catching random clothes